src/main/java/me/botsko/prism/actions/BlockAction.java

```diff
@@ -30,574 +30,534 @@
+                actionData = new CommandActionData();
+                ((CommandActionData) actionData).command = data;
+            }
+        }
+    }
+
+    private BlockActionData getActionData() {
+        return actionData;
+    }
+
+
+    @Override
+    public String getNiceName() {
+        String name = "";
+        BlockActionData blockActionData = getActionData();
+
+        if (blockActionData instanceof SkullActionData) {
+            final SkullActionData ad = (SkullActionData) blockActionData;
+            name += ad.skullType + " ";
+        } else if (blockActionData instanceof SpawnerActionData) {
+            final SpawnerActionData ad = (SpawnerActionData) blockActionData;
+            name += ad.entityType + " ";
+        }
+        name += Prism.getItems().getAlias(getMaterial(), getBlockData());
+        if (blockActionData instanceof SignActionData) {
+            final SignActionData ad = (SignActionData) blockActionData;
+            if (ad.lines != null && ad.lines.length > 0) {
+                name += " (" + TypeUtils.join(ad.lines, ", ") + ")";
+            }
+        } else if (blockActionData instanceof CommandActionData) {
+            final CommandActionData ad = (CommandActionData) blockActionData;
+            name += " (" + ad.command + ")";
+        }
+        if (getActionType().getName().equals("crop-trample") && getMaterial() == Material.AIR) {
+            return "empty soil";
+        }
+        return name;
+    }
+
+    @Override
+    public String getCustomDesc() {
+        if (getActionType().getName().equals("water-bucket") && getBlockData() instanceof Waterlogged) {
+            return "waterlogged";
+        }
+
+        return null;
+    }
+
+    @Override
+    public ChangeResult applyRollback(Player player, QueryParameters parameters, boolean isPreview) {
+        final Block block = getWorld().getBlockAt(getLoc());
+        if (getActionType().doesCreateBlock()) {
+            return removeBlock(player, parameters, isPreview, block);
+        } else {
+            return placeBlock(player, parameters, isPreview, block, false);
+        }
+    }
+
+    @Override
+    public ChangeResult applyRestore(Player player, QueryParameters parameters, boolean isPreview) {
+        final Block block = getWorld().getBlockAt(getLoc());
+        if (getActionType().doesCreateBlock()) {
+            return placeBlock(player, parameters, isPreview, block, false);
+        } else {
+            return removeBlock(player, parameters, isPreview, block);
+        }
+    }
+
+    @Override
+    public ChangeResult applyUndo(Player player, QueryParameters parameters, boolean isPreview) {
+
+        final Block block = getWorld().getBlockAt(getLoc());
+
+        // Undo a drain/ext event (which always remove blocks)
+        // @todo if we ever track rollback/restore for undo, we'll
+        // need logic to do the opposite
+        return placeBlock(player, parameters, isPreview, block, false);
+
+    }
+
+    /**
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Summary javadoc is missing.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5446892)